### DISCUSSION TITLE
Return the old destructor pointer in mem_destructor()

### DISCUSSION THREAD

**Discussion by binarno** [baresip/re#1176] [Ideas/Proposals]
mem_destructor() allows to set a referenced memory destructor.
However, a new destructor may still want to call the previously set destructor.

For instance, a new destructor may just want to perform some other action before continuing the destruction process as before (notifying some other object, logging, etc).

For this reason I propose to add a return value to mem_destructor so that the new destructor can simply continue the previous destruction process:

```c
/**
 * Set or unset a destructor for a memory object
 *
 * @param data Memory object
 * @param dh   called when destroyed, NULL for remove
 * 
 * @return A pointer to the previously set destructor
 */
mem_destroy_h *mem_destructor(void *data, mem_destroy_h *dh)
{
    struct mem *m;

    if (!data)
        return NULL;

    m = get_mem(data);

    MAGIC_CHECK(m);

    mem_destroy_h *old_dh = m->dh;

    m->dh = dh;

    return old_dh;
}
```

**Comment by sreimers** [comment]
I'm not against this, but I wonder how helpful it really is. `mem_destructor` was recently introduced to add a destructor (independent of `mem_alloc`) at a later stage or reset, not really to replace a existing with a new one. You can always `mem_ref` within a destructor to avoid destructing (`free`) if you need multiple stages of destructing. Do you have a real use case for this? And can you show it?

**Comment by binarno** [comment]
In my case I need to extend the destruction process of a call structure with a method that perform some other cleanup and logging.
The library is being use to create a softphone in Qt (C++), and there are some C++ wrappers for the C structures.

Because the call's destructor already knows how to destroy the call structure and ideally is hidden from the client, I would not copy the default call destructor code into the new one, but I would have the new one perform its own things and then call the old destructor so it can destroy the call object.

By modifying mem_destructor to return a pointer to the previous destructor, the new one can do its own things and then simply call the previous destructor to take care of the complete removal of the structure.
